src/lfortran/semantics/ast_symboltable_visitor.cpp

```diff
@@ -1090,6 +1096,57 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
         visit_DeclarationUtil(x);
     }
 
+    void visit_DeclarationPragma(const AST::DeclarationPragma_t &x) {
+        if (x.m_type == AST::LFortranPragma) {
+            std::string t = x.m_text;
+            if (startswith(t, "attributes ")) {
+                t = t.substr(11);
+                if (startswith(t, "simd :: ")) {
+                    t = t.substr(8);
+                    // TODO: for now assume just one variable
+                    // !LF$ attributes simd :: X
+                    std::string var = to_lower(t);
+                    simd_variables.push_back(std::pair(var, x.base.base.loc));
+                } else {
+                    throw SemanticError("Only `simd` attribute supported",
+                        x.base.base.loc);
+                }
+            } else {
+                throw SemanticError("Unsupported LFortran pragma type",
+                    x.base.base.loc);
+            }
+
+        } else {
+            throw SemanticError("The pragma type not supported yet",
+                x.base.base.loc);
+        }
+    }
+
+    void process_simd_variables() {
+        for (auto &var : simd_variables) {
+            ASR::symbol_t *s = current_scope->get_symbol(var.first);
+            if (s == nullptr) {
+                throw SemanticError("The SIMD variable `" + var.first + "` not declared",
+                    var.second);
+            } else {
+                ASR::ttype_t *t = ASRUtils::symbol_type(s);
+                if (ASR::is_a<ASR::Array_t>(*t)) {
+                    ASR::Array_t *a = ASR::down_cast<ASR::Array_t>(t);
+                    a->m_physical_type = ASR::array_physical_typeType::SIMDArray;
+                    // TODO: check all the SIMD requirements here:
+                    // * 1D array
+                    // * the right, compile time, size, compatible type
+                    // * Not allocatable, or pointer
+                } else {
+                    throw SemanticError("The SIMD variable `" + var.first + "` must be an array",
+                        t->base.loc);
+                }
+            }
+
```

Comment:
Minor: I think/guess this logic can be flattened/simplified as follows:

```suggestion
            if (s == nullptr) {
                throw SemanticError("The SIMD variable `" + var.first + "` not declared",
                    var.second);
            }
            ASR::ttype_t *t = ASRUtils::symbol_type(s);
            if ( !ASR::is_a<ASR::Array_t>(*t) ) {
                throw SemanticError("The SIMD variable `" + var.first + "` must be an array", t->base.loc);
            }
            ASR::Array_t *a = ASR::down_cast<ASR::Array_t>(t);
            a->m_physical_type = ASR::array_physical_typeType::SIMDArray;
            // TODO: check all the SIMD requirements here:
            // * 1D array
            // * the right, compile time, size, compatible type
            // * Not allocatable, or pointer
```

Reply:
Actually I think d0f29863f9383dc9d70f3b97f8ac064be66cf6a1 is even better way to fix this.